src/deploy/functions/checkIam.ts

```diff
@@ -104,3 +107,82 @@ export async function checkHttpIam(
   }
   logger.debug("[functions] found setIamPolicy permission, proceeding with deploy");
 }
+
+function reduceEventsToServices(services: Array<Service>, endpoint: backend.Endpoint) {
+  const service = serviceForEndpoint(endpoint);
+  if (!services.find((s) => s.name === service.name)) {
+    services.push(service);
+  }
+  return services;
+}
+
+/**
+ * Checks and sets the roles for specific resource service agents
+ * @param projectId project identifier
+ * @param want backend that we want to deploy
+ * @param have backend that we have currently deployed
+ */
+export async function ensureServiceAgentRoles(
+  projectId: string,
+  want: backend.Backend,
+  have: backend.Backend
+): Promise<void> {
+  // find new services
+  const wantServices = backend.allEndpoints(want).reduce(reduceEventsToServices, []);
+  const haveServices = backend.allEndpoints(have).reduce(reduceEventsToServices, []);
+  const newServices = wantServices.filter(
+    (wantS) => !haveServices.find((haveS) => wantS.name === haveS.name)
+  );
+  if (newServices.length === 0) {
+    return;
+  }
+  // get the full project iam policy
+  let policy: iam.Policy;
+  try {
+    policy = await getIamPolicy(projectId);
+  } catch (err) {
+    logger.warn(
+      "We failed to obtain the IAM policy for the project,",
+      " the function deployment might fail if service ",
+      "accounts do not have the correct assigned roles."
+    );
+    return;
+  }
+  // run in parallel all the missingProjectBindings jobs
+  const findUpdatedBindings: Array<Promise<Array<iam.Binding>>> = [];
+  newServices.forEach((service) =>
+    findUpdatedBindings.push(service.missingProjectBindings(projectId, policy))
+  );
+  const allUpdatedBindings = await Promise.all(findUpdatedBindings);
+  // merge all updated bindings into the policy
+  for (const updatedBindings of allUpdatedBindings) {
+    if (updatedBindings.length === 0) {
+      continue;
+    }
+    updatedBindings.forEach((updatedBinding) => {
+      const ndx = policy.bindings.findIndex(
+        (policyBinding) => policyBinding.role === updatedBinding.role
+      );
+      if (ndx === -1) {
+        policy.bindings.push(updatedBinding);
+        return;
```

Comment:
Is this supposed to be a continue?

Reply:
Nope, this was just returning from the callback. I ended up changing it `for(...)` for better readability in the latest commit.